### ISSUE TITLE
[🐛] Snyk test consumes 12 GB memory and fails with FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory

### ISSUE THREAD

**Issue by s-radyuk** [snyk/snyk#1098]
- `node -v`: v14.0.0
- `npm -v`: 6.14.4
- `snyk -v`: 1.315.0
- OS: macOS Catalina 10.15.3
- Command run: `snyk test `

### Expected behaviour
`snyk test` successfully finishes checking project dependencies.

### Actual behaviour
At some point we started getting error `FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory`. Following this [recomendation](https://support.snyk.io/hc/en-us/articles/360002046418-JavaScript-heap-out-of-memory) we increased allocated memory up to 12 GB but it still didn't help. 

### Steps to reproduce
We have `snyk test` job added to CI pipeline and it was working fine until the end of December 2019. Report was showing 648 dependencies and didn't fail on memory. I agree that our project may be considered large. But snyk test worked without additional memory allocation.

In CI we use linux based docker image and have the same failing behaviour as on macOS. I can include logs from docker machine if needed.

### Debug log from macOS
`$ snyk test --debug
  snyk test <ref *1> { _: [ [Circular *1] ], debug: true } +0ms
| Analyzing gomodules dependencies for ahoy project dir
<--- Last few GCs --->
) [47016:0x10415c000]   307083 ms: Mark-sweep 12191.7 (12206.1) -> 12191.0 (12206.4) MB, 28530.9 / 0.0 ms  (+ 0.0 ms in 13 steps since start of marking, biggest step 0.0 ms, walltime since start of marking 28534 ms) (average mu = 0.107, current mu = 0.000) [47016:0x10415c000]   334996 ms: Mark-sweep 12192.0 (12206.4) -> 12191.3 (12206.6) MB, 27909.4 / 0.0 ms  (+ 0.0 ms in 14 steps since start of marking, biggest step 0.0 ms, walltime since start of marking 27912 ms) (average mu = 0.056, current mu = 0.000)

<--- JS stacktrace --->

`FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x100bc516f node::Abort() (.cold.1) [/usr/local/bin/node]
 2: 0x1000816b9 node::FatalError(char const*, char const*) [/usr/local/bin/node]
 3: 0x100081822 node::OnFatalError(char const*, char const*) [/usr/local/bin/node]
 4: 0x1001809e9 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]
 5: 0x100180993 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]
 6: 0x1002a0c51 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/usr/local/bin/node]
 7: 0x1002a1fb4 v8::internal::Heap::MarkCompactPrologue() [/usr/local/bin/node]
 8: 0x10029f99b v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/usr/local/bin/node]
 9: 0x10029e166 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/local/bin/node]
10: 0x1002a6128 v8::internal::Heap::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/usr/local/bin/node]
11: 0x1002a617e v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/usr/local/bin/node]
12: 0x1002842e5 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [/usr/local/bin/node]
13: 0x1004e79fe v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [/usr/local/bin/node]
14: 0x10074fd99 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit [/usr/local/bin/node]
15: 0x100797f81 Builtins_RegExpPrototypeExec [/usr/local/bin/node]
zsh: abort      snyk test --debug`

**Comment by gitphill** [comment]
Hi 👋 Thanks for raising this. We have seen this with other go module users. We can generate very large dependency trees that cause OOM errors. To try and help we introduced a flag that will prune repeated sub dependencies. Could you try:`snyk test --prune-repeated-subdependencies`

**Comment by s-radyuk** [comment]
Hi @gitphill ! Thank you very much for your suggestion! Unfortunately it didn't work for my project, still failing. Here is the log. 

$ snyk test --prune-repeated-subdependencies --debug
``` 
  snyk test <ref *1> {
  _: [ [Circular *1] ],
  'prune-repeated-subdependencies': true,
  debug: true
} +0ms
| Analyzing gomodules dependencies for ahoy project dir
<--- Last few GCs --->
) [49532:0x10415c000]   364870 ms: Mark-sweep 12191.5 (12206.1) -> 12190.8 (12206.1) MB, 27554.9 / 0.0 ms  (+ 0.0 ms in 15 steps since start of marking, biggest step 0.0 ms, walltime since start of marking 27560 ms) (average mu = 0.054, current mu = 0.000) [49532:0x10415c000]   396147 ms: Mark-sweep 12191.8 (12206.4) -> 12191.1 (12206.4) MB, 31273.0 / 0.0 ms  (+ 0.0 ms in 13 steps since start of marking, biggest step 0.0 ms, walltime since start of marking 31277 ms) (average mu = 0.027, current mu = 0.000)

<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x100bc516f node::Abort() (.cold.1) [/usr/local/bin/node]
 2: 0x1000816b9 node::FatalError(char const*, char const*) [/usr/local/bin/node]
 3: 0x100081822 node::OnFatalError(char const*, char const*) [/usr/local/bin/node]
 4: 0x1001809e9 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]
 5: 0x100180993 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]
 6: 0x1002a0c51 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/usr/local/bin/node]
 7: 0x1002a1fb4 v8::internal::Heap::MarkCompactPrologue() [/usr/local/bin/node]
 8: 0x10029f99b v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/usr/local/bin/node]
 9: 0x10029e166 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/local/bin/node]
10: 0x1002a6128 v8::internal::Heap::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/usr/local/bin/node]
11: 0x1002a617e v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/usr/local/bin/node]
12: 0x1002842e5 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [/usr/local/bin/node]
13: 0x1004e79fe v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [/usr/local/bin/node]
14: 0x10074fd99 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit [/usr/local/bin/node]
15: 0x100797f81 Builtins_RegExpPrototypeExec [/usr/local/bin/node]
zsh: abort      snyk test --prune-repeated-subdependencies --debug

**Comment by anthogez** [comment]
Hi @s-radyuk, we just released Snyk CLI v1.363.0 that contains **go-modules improvements**,
would you mind try it in your environment? I think it should solve your **OutOfMemory** issue, thanks!

**Comment by anthogez** [comment]
We have enough data to prove that this issue is no longer a problem, closing it.
